art/defences/trainer/ibp_certified_trainer_pytorch.py

```diff
@@ -0,0 +1,462 @@
+                else:
+                    certified_loss = certification_loss(interval_preds, y_batch)
+                samples_certified = self._classifier.certify(preds=interval_preds.cpu().detach(), labels=y_batch)
+
+                cert_loss.append(certified_loss)
+                cert_acc.append(np.sum(samples_certified) / batch_size)
+
+                # Concrete PGD loss
+                if self.augment_with_pgd:
+                    i_batch = np.copy(
+                        x_preprocessed[ind[m * self.pgd_params["batch_size"] : (m + 1) * self.pgd_params["batch_size"]]]
+                    ).astype("float32")
+                    o_batch = y_preprocessed[
+                        ind[m * self.pgd_params["batch_size"] : (m + 1) * self.pgd_params["batch_size"]]
+                    ]
+                    # Perform prediction
+                    self.set_forward_mode("attack")
+                    self.attack = ProjectedGradientDescent(
+                        estimator=self._classifier,
+                        eps=self.pgd_params["eps"],
+                        eps_step=self.pgd_params["eps_step"],
+                        max_iter=self.pgd_params["max_iter"],
+                        num_random_init=self.pgd_params["num_random_init"],
+                    )
+                    i_batch = self.attack.generate(i_batch, y=o_batch)
+                    self._classifier.model.zero_grad()
+                else:
+                    i_batch = np.copy(x_preprocessed[ind[m * batch_size : (m + 1) * batch_size]]).astype("float32")
+                    o_batch = y_preprocessed[ind[m * batch_size : (m + 1) * batch_size]]
+                self.set_forward_mode("concrete")
+
+                model_outputs = self._classifier.model.forward(i_batch)
+                acc = self._classifier.get_accuracy(model_outputs, o_batch)
+
+                # Form the loss function
+                non_cert_loss = self._classifier.concrete_loss(
+                    model_outputs, torch.from_numpy(o_batch).to(self._classifier.device)
+                )  # pylint: disable=W0212
+                epoch_non_cert_loss.append(non_cert_loss)
+                non_cert_acc.append(acc)
+
+                loss = certified_loss * loss_weighting_k + non_cert_loss * (1 - loss_weighting_k)
+                # Do training
+                if self._classifier._use_amp:  # pragma: no cover # pylint: disable=W0212
+                    from apex import amp  # pylint: disable=E0611
+
+                    with amp.scale_loss(loss, self._classifier._optimizer) as scaled_loss:  # pylint: disable=W0212
+                        scaled_loss.backward()
+
+                else:
+                    loss.backward()
+
+                self._classifier._optimizer.step()  # pylint: disable=W0212
+                self._classifier.re_convert()
+
+                if verbose:
+                    pbar.set_description(
+                        f"Bound {bound:.2f}: "  # pylint: disable=W0212
```

Comment:
Is this in the correct line?
```suggestion
                        f"Bound {bound:.2f}: "
```

Reply:
Hi Beat, we removed the need for this warning, but yes oddly enough this was the line despite the warning actually occurring further down. It seemed to be that it needed to be placed when the string was first opened .